### PR TITLE
Move `ScalarValue` tests alongside implementation, move `from_slice` to `datafusion_core`

### PR DESCRIPTION
**Pull request by alamb**
# Which issue does this PR close?
Closes https://github.com/apache/arrow-datafusion/issues/2913

 # Rationale for this change
I am working on https://github.com/apache/arrow-datafusion/pull/2891 and it was not clear there were existing `ScalarValue` tests because they were in a different module than their implementation

Implementation: https://github.com/apache/arrow-datafusion/blob/fd64e6f2e5ca5ba8c33c431b75a3ba9441091970/datafusion/common/src/scalar.rs
Tests https://github.com/apache/arrow-datafusion/blob/85ca8be669dc1fbba4b2eefbf99fc2ed1546cc0c/datafusion/core/src/scalar.rs

🤯 

# What changes are included in this PR?
1. Move `ScalarValue` tests alongside implementation
2. move `from_slice` to `datafusion_core` (as it is used in the ScalarValue implementation)

# Are there any user-facing changes?
`from_slice` is now in the `datafusion_core` crate (used to be in `datafusion_physical_expr` crate), but their is a `pub use` so no existing code needs to be changed (they can still import `from_slice` from `datafusion_physical_expr`)

### CODE REVIEW

File: datafusion/common/src/scalar.rs [hunk 1]
```diff
@@ -1987,3 +1987,1176 @@ impl ScalarType<i64> for TimestampNanosecondType {
         ScalarValue::TimestampNanosecond(r, None)
     }
 }
+
```
Comment (alamb) [inline datafusion/common/src/scalar.rs:1990]: these tests are moved, verbatim. The only change is that this line

```rust
    use arrow::{array::*, datatypes::*};
```

was removed at the direction of the rust compiler as it was no longer used

**Review by alamb — COMMENTED** [review]
Github makes this look like a massive diff, but it has about 5 lines of substance only

**Review by andygrove — APPROVED** [review]
Thanks @alamb. The current state was likely a result of the recent refactoring of crates and modules.

**Comment by alamb** [comment]
cc @comphead  this may affect https://github.com/apache/arrow-datafusion/pull/2893

Would you like me to wait for that to be ready to merge this PR?

**Comment by alamb** [comment]
@comphead  I merged this PR to make progress on other PRs. I will be happy to help rebase https://github.com/apache/arrow-datafusion/pull/2893 if you have trouble after this.

**Pull request merged**